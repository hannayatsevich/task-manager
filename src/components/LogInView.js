import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {Redirect} from 'react-router-dom';
import {switchUserStatusAC} from '../redux/actionCreators';
import {escapeHTML, validateLogInData} from './_functionsAndConstants';

import '../styles/LogInView.css';

class LogInView extends React.PureComponent{
  static propTypes = {
    users: PropTypes.arrayOf(PropTypes.object).isRequired,//redux - state.taskManagerUsersData.users
    isLoggedIn: PropTypes.object.isRequired,//redux - state.taskManagerUsersData.isLoggedIn, //логин, пароль пользователя или null, null
    lastVisitedUrl: PropTypes.string.isRequired,//redux - state.taskManagerUsersData.lastVisitedUrl
  };
  constructor(props) {
    super(props);

    this.nameInputRef = React.createRef();
    this.pswInputRef = React.createRef();

    this.state = {  
      nameInputValue: '',
      nameInputError: null,
      pswInputValue: '',
      pswInputError: null,
      isValidated: false,
      redirect: false,
    };
  };

  validateInputs = () => {
    let nameValue = escapeHTML(this.nameInputRef.current.value);
    let pswValue = escapeHTML(this.pswInputRef.current.value);
    let validationInfo = validateLogInData(this.props.users, nameValue, pswValue);
    this.setState({nameInputValue: nameValue, pswInputValue: pswValue, isValidated: validationInfo.isValidated, nameInputError: validationInfo.nameInputError, pswInputError: validationInfo.pswInputError})
  };
  logInUser = () => {    
    this.props.dispatch(switchUserStatusAC(this.state.nameInputValue, this.state.pswInputValue));
    this.setState({redirect: true});
  };
  cancelLogInUser = () => {
    this.setState({redirect: true});
  };

  render() {
    //console.log('LogInView render');
    if(this.state.redirect)
      return <Redirect to = {this.props.lastVisitedUrl}/>;
    else
      return (        
        <div className = 'LogInView'>
          <div className = 'LogInWindow'>
            <p>User Name</p>
            <p><input type = 'text' autoFocus value = {this.state.nameInputValue} onChange = {this.validateInputs} ref = {this.nameInputRef}/></p>
            <p className = 'ErrorMsg'>{this.state.nameInputError && <span>{this.state.nameInputError}</span>}</p>
            <p>User Password</p>
            <p><input type = 'text' value = {this.state.pswInputValue} onChange = {this.validateInputs} ref = {this.pswInputRef}/></p>
            <p className = 'ErrorMsg'>{this.state.pswInputError && <span>{this.state.pswInputError}</span>}</p>
            <p>
              <input type = 'button' value = 'Log in' disabled = {!this.state.isValidated} onClick = {this.logInUser}/>
              <input type = 'button' value = 'Cancel' onClick = {this.cancelLogInUser}/>
            </p>
          </div>       
        </div>        
      );
    };
};

const mapStateToProps = function (state) {
  return {
    users: state.taskManagerUsersData.users,
    isLoggedIn: state.taskManagerUsersData.isLoggedIn,
    lastVisitedUrl: state.taskManagerUsersData.lastVisitedUrl,
 };
};

export {LogInView};
export default connect(mapStateToProps)(LogInView);
