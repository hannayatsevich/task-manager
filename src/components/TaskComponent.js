import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {getDateString, getActionDataObject, escapeHTML, validateTaskData, getDateFromString, getInitialFormData} from './_functionsAndConstants';
import {DELETE_TASK, EDIT_TASK, getUsersDataThunkForActionAC} from '../redux/actionCreators';
import '../styles/TaskComponent.css';

class TaskComponent extends React.Component{
  static propTypes = {
    task:PropTypes.shape({
      taskID: PropTypes.number.isRequired,
      taskDate: PropTypes.string.isRequired,
      taskName: PropTypes.string.isRequired,
      taskBody: PropTypes.string.isRequired,
      taskTags: PropTypes.string,
    }).isRequired,
    needDate: PropTypes.bool.isRequired,//для DayView false, для AllTasksView true
    userName: PropTypes.string.isRequired,
  };
  constructor(props) {
    super(props);
    let taskDate = getDateString(new Date(props.task.taskDate));
    
    this.taskDateRef = React.createRef();
    this.taskNameRef = React.createRef();
    this.taskBodyRef = React.createRef();
    this.taskTagsRef = React.createRef();

    this.state = {    
      taskDate: taskDate,  
      tagsArray: props.task.taskTags.split(','),
      isEditable: false,     
      isAnimated: false,
      isValidated: true,
      formData: getInitialFormData(taskDate, props.task.taskName, props.task.taskBody, props.task.taskTags),
    };
  };

  animateTask = () => {
    this.setState({isAnimated: true});
  };
  deleteTask = () => {
    let actionDataObject = getActionDataObject(DELETE_TASK, {userName: this.props.userName, tasksArray: [this.props.task]});
    this.props.dispatch(getUsersDataThunkForActionAC(this.props.dispatch, actionDataObject));
  };
  setIsEditable = () => {
    this.setState({isEditable: true});
  };
  validateInputs = () => {
    let taskDate = escapeHTML(this.taskDateRef.current.value);
    let taskName = escapeHTML(this.taskNameRef.current.value);
    let taskBody = escapeHTML(this.taskBodyRef.current.value);
    let taskTags = escapeHTML(this.taskTagsRef.current.value);
    let newFormData = {...this.state.formData, 
                      dateInputValue: taskDate, 
                      taskNameInputValue: taskName, 
                      taskBodyInputValue: taskBody, 
                      taskTagsInputValue: taskTags};
    let validationData = validateTaskData(newFormData);
    newFormData.dateInputError = validationData.errorTexts.dateErrorMessage;
    newFormData.taskNameInputError = validationData.errorTexts.taskNameErrorMessage;
    newFormData.taskBodyInputError = validationData.errorTexts.taskBodyErrorMessage;
    newFormData.taskTagsInputError = validationData.errorTexts.taskTagsErrorMessage;
    this.setState({isValidated: validationData.isValidated, formData: newFormData});
  };
  editTask = () => {
    let taskDate = getDateFromString(this.state.formData.dateInputValue);
    if ( this.props.task.taskDate !== taskDate 
      || this.props.task.taskName !== this.state.formData.taskBodyInputValue 
      || this.props.task.taskBody !== this.state.formData.taskBodyInputValue 
      || this.props.task.taskTags !== this.state.formData.taskTagsInputValue) {
        let task = {...this.props.task, taskName: this.state.formData.taskNameInputValue, taskBody: this.state.formData.taskBodyInputValue, taskTags: this.state.formData.taskTagsInputValue, taskDate: taskDate};
        let actionDataObject = getActionDataObject(EDIT_TASK, {userName: this.props.userName, taskObject: task});
        this.props.dispatch(getUsersDataThunkForActionAC(this.props.dispatch, actionDataObject));
        this.setState({isEditable: false});
    }
    else
      this.canceEditingTask();
  };
  canceEditingTask = () => {
    let taskDate = getDateString(new Date(this.props.task.taskDate));
    let formData = getInitialFormData(taskDate, this.props.task.taskName, this.props.task.taskBody, this.props.task.taskTags);
    this.setState({isEditable: false, formData: formData});
  };
  UNSAFE_componentWillReceiveProps(nextProps) {
    let taskDate = getDateString(new Date(nextProps.task.taskDate));
    let tagsArray = nextProps.task.taskTags.split(',');
    let formData = getInitialFormData(taskDate, nextProps.task.taskName, nextProps.task.taskBody, nextProps.task.taskTags);
    this.setState({taskDate: taskDate, tagsArray: tagsArray, formData: formData});
  };
  componentDidUpdate() {
    if(this.state.isAnimated)
      setTimeout(this.deleteTask, 800)
  };
  render() {
    //console.log('TaskComponent render');
    if (this.state.isEditable)
      return (
        <div className = 'EditTaskWindow'>
          <p><label>Task date: <input type = 'text' ref = {this.taskDateRef} value = {this.state.formData.dateInputValue} onChange = {this.validateInputs}/></label></p>
          <p className = 'ErrorMsg'>{this.state.formData.dateInputError && <span>{this.state.formData.dateInputError}</span>}</p>
          <p><label>Task name: <input type = 'text' ref = {this.taskNameRef} value = {this.state.formData.taskNameInputValue} onChange = {this.validateInputs}/></label></p>
          <p className = 'ErrorMsg'>{this.state.formData.taskNameInputError && <span>{this.state.formData.taskNameInputError}</span>}</p>
          <p><label>Task body: <textarea ref = {this.taskBodyRef} value = {this.state.formData.taskBodyInputValue} onChange = {this.validateInputs}/></label></p>
          <p className = 'ErrorMsg'>{this.state.formData.taskBodyInputError && <span>{this.state.formData.taskBodyInputError}</span>}</p>
          <p><label>Task tags: <input type = 'text' ref = {this.taskTagsRef} value = {this.state.formData.taskTagsInputValue} onChange = {this.validateInputs}/></label></p>
          <p className = 'ErrorMsg'>{this.state.formData.taskTagsInputError && <span>{this.state.formData.taskTagsInputError}</span>}</p>
          <p>
            <input type = 'button' value = 'Edit task' disabled = {!this.state.isValidated} onClick = {this.editTask}/>
            <input type = 'button' value = 'Cancel' onClick = {this.canceEditingTask}/>
          </p>

        </div>
      );
    else 
      return (
        <div className = 'TaskComponent' id = {this.state.isAnimated ? 'TaskComponentAnimate' : undefined}>
          {this.props.needDate && <p className = 'TaskDate'>{getDateString(new Date(this.props.task.taskDate))}</p>}
          <p className = 'TaskName'>{this.props.task.taskName}</p>
          <p className = 'TaskBody'>{this.props.task.taskBody}</p>
          <p className = 'TaskTags'>{this.state.tagsArray && this.state.tagsArray.map((element, index) => <span className = 'TaskTag' key = {index}>{element}</span>)}</p>
          <p className = 'TaskBtns'>
            <input type = 'button' value = 'Delete' onClick = {this.animateTask}></input>
            <input type = 'button' value = 'Edit' onClick = {this.setIsEditable}></input>
          </p>
        </div>
    );
  };  
};
export {TaskComponent};
export default connect()(TaskComponent);