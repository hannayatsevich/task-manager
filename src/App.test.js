
import React from 'react';
import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { TaskComponent } from './components/TaskComponent';
import ConnectedLogInView from './components/LogInView';
import {LogInView} from './components/LogInView';
import { getNumbersFromDate, getDateString, getDateFromString, getInitialFormData, 
        validateTaskData, getActionDataObject, escapeHTML, validateLogInData } from './components/_functionsAndConstants';
import {deleteTaskAC, editTaskAC, switchUserStatusAC, DELETE_TASK, EDIT_TASK, SWITCH_USER_STATUS} from './redux/actionCreators';
import taskManagerUsersDataReducer from './redux/taskManagerUsersDataReducer';
        
let data = require('./data.json');

describe('TaskComponent functions', () => {
  it('check getNumbersFromDate', () => {
    let inpit1 = new Date(2019, 1, 1);
    let output1 = {yearNum: 2019, monthNum: 2, dayNum: 1};
    let inpit2 = new Date(1990, 3, 27);
    let output2 = {yearNum: 1990, monthNum: 4, dayNum: 27};
      
    expect(getNumbersFromDate(inpit1)).toEqual(output1);
    expect(getNumbersFromDate(inpit2)).toEqual(output2);
  });
  it('check getDateString', () => {
    let inpit1 = new Date(2019, 1, 1);
    let output1 = '01.02.2019';
    let inpit2 = new Date(1990, 3, 27);
    let output2 = '27.04.1990';
      
    expect(getDateString(inpit1)).toEqual(output1);
    expect(getDateString(inpit2)).toEqual(output2);
  });
  it('check getDateFromString', () => {
    let taskDate1 = '01.02.2019';
    let taskDate2 = '05.07.2020';
    let taskDateOutput1 = new Date(2019, 1, 1);
    let taskDateOutput2 = new Date(2020, 6, 5);

    expect(getDateFromString(taskDate1)).toEqual(taskDateOutput1);
    expect(getDateFromString(taskDate2)).toEqual(taskDateOutput2);
  });
  it('check escapeHTML', () => {
    let text1 = null;
    let text2 = 'text';
    let text3 = '&text';
    let text4 = '<text';
    let text5 = '>text';
    let text6 = '"text';
    let text7 = "'text";
    let output1 = null;
    let output2 = 'text';
    let output3 = '&amp;text';
    let output4 = '&lt;text';
    let output5 = '&gt;text';
    let output6 = '&quot;text';
    let output7 = "&#039;text";

    expect(escapeHTML(text1)).toBe(output1);
    expect(escapeHTML(text2)).toBe(output2);
    expect(escapeHTML(text3)).toBe(output3);
    expect(escapeHTML(text4)).toBe(output4);
    expect(escapeHTML(text5)).toBe(output5);
    expect(escapeHTML(text6)).toBe(output6);
    expect(escapeHTML(text7)).toBe(output7);
  });
  it('check getInitialFormData', () => {
    let taskDate1 = '01/02/2019';
    //let taskDate2 = '05.07.2020';
    let taskDate3 = '11-04-1900';
    let taskDate4 = '31.09.1750';
    let taskDate5 = '07.02.2050';
    let taskName1 = 'Task1#';
    let taskName2 = 'Task2';
    //let taskName3 = 'Task3';
    let taskName4 = 'Task4';
    let taskName5 = 'Task5';
    let taskBody1 = 'TaskBody1';
    let taskBody2 = 'TaskBody2';
    let taskBody3 = 'TaskBody3';
    //let taskBody4 = 'TaskBody4';
    let taskBody5 = 'TaskBody5';
    let taskTags1 = 'Tag1, Tag2, Tag3';
    let taskTags2 = 'Tag4, Tag5, Tag6';
    let taskTags3 = 'Tag7, Tag8, Tag9';
    let taskTags4 = 'Tag10-Tag11-Tag12';
    //let taskTags5 = 'Tag13, Tag14, Tag15';
    let formData1 = {
        dateInputValue: taskDate1,
        dateInputError: null,
        taskNameInputValue: taskName1,
        taskNameInputError: null,
        taskBodyInputValue: taskBody1,
        taskBodyInputError: null,
        taskTagsInputValue: taskTags1,
        taskTagsInputError: null,
      };
      let formData2 = {
        dateInputValue: '',
        dateInputError: null,
        taskNameInputValue: taskName2,
        taskNameInputError: null,
        taskBodyInputValue: taskBody2,
        taskBodyInputError: null,
        taskTagsInputValue: taskTags2,
        taskTagsInputError: null,
      };
      let formData3 = {
        dateInputValue: taskDate3,
        dateInputError: null,
        taskNameInputValue: '',
        taskNameInputError: null,
        taskBodyInputValue: taskBody3,
        taskBodyInputError: null,
        taskTagsInputValue: taskTags3,
        taskTagsInputError: null,
      };
      let formData4 = {
        dateInputValue: taskDate4,
        dateInputError: null,
        taskNameInputValue: taskName4,
        taskNameInputError: null,
        taskBodyInputValue: '',
        taskBodyInputError: null,
        taskTagsInputValue: taskTags4,
        taskTagsInputError: null,
      };
      let formData5 = {
        dateInputValue: taskDate5,
        dateInputError: null,
        taskNameInputValue: taskName5,
        taskNameInputError: null,
        taskBodyInputValue: taskBody5,
        taskBodyInputError: null,
        taskTagsInputValue: '',
        taskTagsInputError: null,
      };

      expect(getInitialFormData(taskDate1, taskName1, taskBody1, taskTags1)).toEqual(formData1);
      expect(getInitialFormData('', taskName2, taskBody2, taskTags2)).toEqual(formData2);
      expect(getInitialFormData(taskDate3, '', taskBody3, taskTags3)).toEqual(formData3);
      expect(getInitialFormData(taskDate4, taskName4, '', taskTags4)).toEqual(formData4);
      expect(getInitialFormData(taskDate5, taskName5, taskBody5, '')).toEqual(formData5);
  });
  it('check validateTaskData', () => {
    let taskDate1 = '01/02/2019';
    //let taskDate2 = '05.07.2020';
    let taskDate3 = '11-04-1900';
    let taskDate4 = '31.09.1750';
    let taskDate5 = '07.02.2050';
    let taskName1 = 'Task1#';
    let taskName2 = 'Task2';
    //let taskName3 = 'Task3';
    let taskName4 = 'Task4';
    let taskName5 = 'Task5';
    let taskBody1 = 'TaskBody1';
    let taskBody2 = 'TaskBody2';
    let taskBody3 = 'TaskBody3';
    //let taskBody4 = 'TaskBody4';
    let taskBody5 = 'TaskBody5';
    let taskTags1 = 'Tag1, Tag2, Tag3';
    let taskTags2 = 'Tag4, Tag5, Tag6';
    let taskTags3 = 'Tag7, Tag8, Tag9';
    let taskTags4 = 'Tag10-Tag11-Tag12';
    //let taskTags5 = 'Tag13, Tag14, Tag15';
    let formData1 = {
        dateInputValue: taskDate1,
        dateInputError: null,
        taskNameInputValue: taskName1,
        taskNameInputError: null,
        taskBodyInputValue: taskBody1,
        taskBodyInputError: null,
        taskTagsInputValue: taskTags1,
        taskTagsInputError: null,
      };
      let formData2 = {
        dateInputValue: '',
        dateInputError: null,
        taskNameInputValue: taskName2,
        taskNameInputError: null,
        taskBodyInputValue: taskBody2,
        taskBodyInputError: null,
        taskTagsInputValue: taskTags2,
        taskTagsInputError: null,
      };
      let formData3 = {
        dateInputValue: taskDate3,
        dateInputError: null,
        taskNameInputValue: '',
        taskNameInputError: null,
        taskBodyInputValue: taskBody3,
        taskBodyInputError: null,
        taskTagsInputValue: taskTags3,
        taskTagsInputError: null,
      };
      let formData4 = {
        dateInputValue: taskDate4,
        dateInputError: null,
        taskNameInputValue: taskName4,
        taskNameInputError: null,
        taskBodyInputValue: '',
        taskBodyInputError: null,
        taskTagsInputValue: taskTags4,
        taskTagsInputError: null,
      };
      let formData5 = {
        dateInputValue: taskDate5,
        dateInputError: null,
        taskNameInputValue: taskName5,
        taskNameInputError: null,
        taskBodyInputValue: taskBody5,
        taskBodyInputError: null,
        taskTagsInputValue: '',
        taskTagsInputError: null,
      };

      let validationData1 = {isValidated: false, 
              errorTexts: {
                dateErrorMessage: 'Please, enter date in "xx.xx.xxxx" format',
                taskNameErrorMessage: 'Task name must consist of letters, numbers and symbols -./\\?',
                taskBodyErrorMessage: null,
                taskTagsErrorMessage: null,
              },
            };
      let validationData2 = {isValidated: false, 
              errorTexts: {
                dateErrorMessage: 'Task date field must be filled',
                taskNameErrorMessage: null,
                taskBodyErrorMessage: null,
                taskTagsErrorMessage: null,
              },
            };
      let validationData3 = {isValidated: false, 
              errorTexts: {
                dateErrorMessage: 'Please, enter date in "xx.xx.xxxx" format',
                taskNameErrorMessage: 'Task name field must be filled',
                taskBodyErrorMessage: null,
                taskTagsErrorMessage: null,
              },
            };
      let validationData4 = {isValidated: false, 
              errorTexts: {
                dateErrorMessage: "Please, enter year between 2000 and 2050",
                taskNameErrorMessage: null,
                taskBodyErrorMessage: 'Task body field must be filled',
                taskTagsErrorMessage: 'Please, enter comma-saparated list of no more than 5 tags',
              },
            };
      let validationData5 = {isValidated: true, 
              errorTexts: {
                dateErrorMessage: null,
                taskNameErrorMessage: null,
                taskBodyErrorMessage: null,
                taskTagsErrorMessage: null,
              },
            };	

      expect(validateTaskData(formData1)).toEqual(validationData1);
      expect(validateTaskData(formData2)).toEqual(validationData2);
      expect(validateTaskData(formData3)).toEqual(validationData3);
      expect(validateTaskData(formData4)).toEqual(validationData4);
      expect(validateTaskData(formData5)).toEqual(validationData5);
  });
  it('check getActionDataObject', () => {
    let action1 = 'EDIT_TASK';
    let action2 = 'DELETE_TASK';
    let data1 = {userName: data[0].userName, taskObject: data[0].userTasks[0]};
    let data2 = {userName: data[1].userName, tasksArray: [data[1].userTasks[1]]};
    
    let output1 = {
      action: 'EDIT_TASK',
      data: {userName: "OliviaSmith", taskObject: {"taskID": 1, "taskDate": "2020-02-02T21:00:00.000Z", "taskName": "Smth1", "taskBody": "Do smth1","taskTags": "do,smth,task"}},
    };
    let output2 = {
      action: 'DELETE_TASK',
      data: {userName: "HenryOscar", tasksArray: [{"taskID": 2, "taskDate": "2020-10-31T21:00:00.000Z", "taskName": "Smth21", "taskBody": "Do smth21","taskTags": "do,task"}]},
    };
      
    expect(getActionDataObject(action1, data1)).toEqual(output1);
    expect(getActionDataObject(action2, data2)).toEqual(output2);
  });
});
describe('TaskComponent', () => {  
  let taskComponent;

  it('component task snapshot, edit task snapshot', () => {  
    let taskComponent = renderer.create(<TaskComponent task = {data[0].userTasks[0]} userName = {data[0].userName} needDate = {true}/>);
    let taskComponentTree = taskComponent.toJSON()
    expect(taskComponentTree).toMatchSnapshot();
    const editTask = taskComponent.root.find( el => el.type === 'input'&& el.props.value === 'Edit');
    editTask.props.onClick();
    taskComponentTree = taskComponent.toJSON()
    expect(taskComponentTree).toMatchSnapshot();
  });  
  beforeEach(()=> {
    taskComponent = mount(<TaskComponent task = {data[0].userTasks[0]} userName = {data[0].userName} needDate = {true}/>);
  })
  it('component exists, check props', () => {  
    expect(taskComponent).toBeTruthy();
    expect(taskComponent.props().task).toEqual(data[0].userTasks[0]);
    expect(taskComponent.props().userName).toEqual(data[0].userName);
    expect(taskComponent.props().needDate).toEqual(true);
  });
  it('check nodes inner texts from props', () => {  
    expect(taskComponent.find('.TaskDate').text()).toEqual(getDateString(new Date(data[0].userTasks[0].taskDate)));
    expect(taskComponent.find('.TaskName').text()).toEqual(data[0].userTasks[0].taskName);
    expect(taskComponent.find('.TaskBody').text()).toEqual(data[0].userTasks[0].taskBody);
    expect(taskComponent.find('.TaskTags').text()).toEqual(data[0].userTasks[0].taskTags.split(',').join(''));
  
  });
  it('simulate "Delete" click', () => {
    let handleDeleteClick = jest.fn();
    let deleteInput = taskComponent.find('input').at(0);  
    deleteInput.props().onClick = handleDeleteClick;  
    deleteInput.props().onClick();
    expect(handleDeleteClick).toHaveBeenCalledTimes(1);  
  });
  it('simulate "Edit" click, check nodes values, change node, simulate "Edit task" click, simulate "Cancel" click', () => {    
    taskComponent.find('input').at(1).simulate('click'); 

    expect(taskComponent.find('input').at(0).props().value).toEqual(getDateString(new Date(data[0].userTasks[0].taskDate)));
    expect(taskComponent.find('input').at(1).props().value).toEqual(data[0].userTasks[0].taskName);
    expect(taskComponent.find('textarea').at(0).props().value).toEqual(data[0].userTasks[0].taskBody);
    expect(taskComponent.find('input').at(2).props().value).toEqual(data[0].userTasks[0].taskTags);

    taskComponent.find('input').at(1).props().value = 'x'; 
    expect(taskComponent.find('input').at(1).props().value).toEqual('x'); 

    let handleEditTaskClick = jest.fn();
    let editTaskInput = taskComponent.find('input').at(3);
    editTaskInput.props().onClick = handleEditTaskClick;
    editTaskInput.props().onClick();
    expect(handleEditTaskClick).toHaveBeenCalledTimes(1);

    let handleCancelClick = jest.fn();
    let cancelTaskInput = taskComponent.find('input').at(4);
    cancelTaskInput.props().onClick = handleCancelClick;
    cancelTaskInput.props().onClick();
    expect(handleCancelClick).toHaveBeenCalledTimes(1);
  });
});
describe('TaskComponent actionCreators', () => {    
  it('check deleteTaskAC', () => {  
    expect(deleteTaskAC(data)).toEqual({type: DELETE_TASK, usersData: data});
  }); 
  it('check editTaskAC', () => {  
    expect(editTaskAC(data)).toEqual({type: EDIT_TASK, usersData: data});
  });  
});
describe('TaskComponent reducer', () => {    
  it('check reducer on delete task', () => {
    let data1 = [data[0]];
    let data2 = [JSON.parse(JSON.stringify(data[0]))];
    const state = {
      users: data,//массив хэшей-данных о users
      mode: 1,//1 - загрузка, 2 - ошибка, 3 - данные получены
      isLoggedIn: {userName: null, userPassword: null},//логин, пароль пользователя или null, null
      lastVisitedUrl: '/',
    };
    expect(taskManagerUsersDataReducer(state, {type: DELETE_TASK, usersData: data1})).toEqual({...state, users: data2});
  }); 
  it('check reducer on edit task', () => {
    let data1 = [data[0]];
    let data2 = [JSON.parse(JSON.stringify(data[0]))];
    const state = {
      users: data,//массив хэшей-данных о users
      mode: 1,//1 - загрузка, 2 - ошибка, 3 - данные получены
      isLoggedIn: {userName: null, userPassword: null},//логин, пароль пользователя или null, null
      lastVisitedUrl: '/',
    };
    expect(taskManagerUsersDataReducer(state, {type: EDIT_TASK, usersData: data1})).toEqual({...state, users: data2});
  });    
});
describe('LogInView functions', () => {
    it('check validateLogInData', () => {  
      let userName1 = 'OliviaSmith1';
      let userName2 = 'HenryOscar';
      let userPsw1 = 'something';
      let userPsw2 = 'somethingelse';
      let userPsw3 = 'somethingelse1';
      let output1 = {isValidated: false, 
                    nameInputError: 'There is no User with this name', 
                    pswInputError: 'Password is invalid, please, try one more time',
                  };
      let output2 = {isValidated: false, 
                    nameInputError: 'User Name field must be filled', 
                    pswInputError: 'Password is invalid, please, try one more time',
                  };
      let output3 = {isValidated: false, 
                    nameInputError: null, 
                    pswInputError: 'Password is invalid, please, try one more time',
                  };
      let output4 = {isValidated: false, 
                    nameInputError: null, 
                    pswInputError: 'Password field must be filled'
                  };
      let output5 = {isValidated: true, 
                    nameInputError: null, 
                    pswInputError: null
                  };
      
      expect(validateLogInData(data, userName1, userPsw1)).toEqual(output1);
      expect(validateLogInData(data, '', userPsw1)).toEqual(output2);
      expect(validateLogInData(data, userName2, userPsw3)).toEqual(output3);
      expect(validateLogInData(data, userName2, '')).toEqual(output4);
      expect(validateLogInData(data, userName2, userPsw2)).toEqual(output5);
  });
});
describe('LogInView + redux', () => {
  const mockStore = configureStore();
  const initialState = {
    taskManagerUsersData: {
      users: [...data],
      isLoggedIn: {userName: data[0].userName, userPassword: data[0].userPassword},
      lastVisitedUrl: '/',
    }
  };
  let store;
  let connectedLogInView;  
  it('render the connected component, snapshot', () => {
    connectedLogInView = renderer.create(<Provider store = {store}><ConnectedLogInView/></Provider>).toJSON();
    expect(connectedLogInView).toMatchSnapshot();
  });
  beforeEach(()=> {
    store = mockStore(initialState);
    connectedLogInView = mount(<Provider store = {store}><ConnectedLogInView/></Provider>);
  });
  it('check props match with mock redux state', () => {
    expect(connectedLogInView.find(LogInView).props().users).toEqual(initialState.taskManagerUsersData.users);
    expect(connectedLogInView.find(LogInView).props().isLoggedIn).toEqual(initialState.taskManagerUsersData.isLoggedIn);
    expect(connectedLogInView.find(LogInView).props().lastVisitedUrl).toEqual(initialState.taskManagerUsersData.lastVisitedUrl);
  });
  it('simulate "Log in" click', () => {     
    let handleLogInClick = jest.fn();
    let logInInput = connectedLogInView.find(LogInView).find('input').at(2);  
    logInInput.props().onClick = handleLogInClick;  
    logInInput.props().onClick();
    expect(handleLogInClick).toHaveBeenCalledTimes(1);      
  });
  it('simulate "Cancel" click', () => {     
    let handleCancelClick = jest.fn();
    let logInInput = connectedLogInView.find(LogInView).find('input').at(3);  
    logInInput.props().onClick = handleCancelClick;  
    logInInput.props().onClick();
    expect(handleCancelClick).toHaveBeenCalledTimes(1);      
  });
});
describe('LogInView actionCreators', () => {    
  it('check switchUserStatusAC', () => {  
    expect(switchUserStatusAC(data[0].userName, data[0].userPassword)).toEqual({type: SWITCH_USER_STATUS, userStatus: {userName: data[0].userName, userPassword: data[0].userPassword}});
  }); 
});
describe('LogInView reducer', () => {    
  it('check reducer on switch user status', () => {
    const state = {
      users: data,//массив хэшей-данных о users
      mode: 1,//1 - загрузка, 2 - ошибка, 3 - данные получены
      isLoggedIn: {userName: null, userPassword: null},//логин, пароль пользователя или null, null
      lastVisitedUrl: '/',
    };
    expect(taskManagerUsersDataReducer(state, {type: SWITCH_USER_STATUS, userStatus: {userName: 'Henry', userPassword: 'xxx'}})).toEqual({...state, isLoggedIn: {userName: 'Henry', userPassword: 'xxx'}});
  });  
});
