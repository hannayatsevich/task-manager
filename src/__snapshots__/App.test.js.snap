// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogInView + redux render the connected component, snapshot 1`] = `
<div
  className="LogInView"
>
  <div
    className="LogInWindow"
  >
    <p>
      User Name
    </p>
    <p>
      <input
        autoFocus={true}
        onChange={[Function]}
        type="text"
        value=""
      />
    </p>
    <p
      className="ErrorMsg"
    />
    <p>
      User Password
    </p>
    <p>
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
    </p>
    <p
      className="ErrorMsg"
    />
    <p>
      <input
        disabled={true}
        onClick={[Function]}
        type="button"
        value="Log in"
      />
      <input
        onClick={[Function]}
        type="button"
        value="Cancel"
      />
    </p>
  </div>
</div>
`;

exports[`TaskComponent component task snapshot, edit task snapshot 1`] = `
<div
  className="TaskComponent"
>
  <p
    className="TaskDate"
  >
    03.02.2020
  </p>
  <p
    className="TaskName"
  >
    Smth1
  </p>
  <p
    className="TaskBody"
  >
    Do smth1
  </p>
  <p
    className="TaskTags"
  >
    <span
      className="TaskTag"
    >
      do
    </span>
    <span
      className="TaskTag"
    >
      smth
    </span>
    <span
      className="TaskTag"
    >
      task
    </span>
  </p>
  <p
    className="TaskBtns"
  >
    <input
      onClick={[Function]}
      type="button"
      value="Delete"
    />
    <input
      onClick={[Function]}
      type="button"
      value="Edit"
    />
  </p>
</div>
`;

exports[`TaskComponent component task snapshot, edit task snapshot 2`] = `
<div
  className="EditTaskWindow"
>
  <p>
    <label>
      Task date: 
      <input
        onChange={[Function]}
        type="text"
        value="03.02.2020"
      />
    </label>
  </p>
  <p
    className="ErrorMsg"
  />
  <p>
    <label>
      Task name: 
      <input
        onChange={[Function]}
        type="text"
        value="Smth1"
      />
    </label>
  </p>
  <p
    className="ErrorMsg"
  />
  <p>
    <label>
      Task body: 
      <textarea
        onChange={[Function]}
        value="Do smth1"
      />
    </label>
  </p>
  <p
    className="ErrorMsg"
  />
  <p>
    <label>
      Task tags: 
      <input
        onChange={[Function]}
        type="text"
        value="do,smth,task"
      />
    </label>
  </p>
  <p
    className="ErrorMsg"
  />
  <p>
    <input
      disabled={false}
      onClick={[Function]}
      type="button"
      value="Edit task"
    />
    <input
      onClick={[Function]}
      type="button"
      value="Cancel"
    />
  </p>
</div>
`;
